osquery/filesystem/windows/fileops.cpp

```diff
@@ -227,6 +231,129 @@ static DWORD getNewAclSize(PACL dacl,
+      ::GetCurrentProcess(),
+      TOKEN_IMPERSONATE | TOKEN_QUERY | TOKEN_DUPLICATE | STANDARD_RIGHTS_READ,
+      &process_token);
+  if (!status) {
+    return Status(-1, "OpenProcessToken failed");
+  }
+
+  status = ::DuplicateToken(
+      process_token, SecurityImpersonation, &impersonate_token);
+  ::CloseHandle(process_token);
+
+  if (!status) {
+    return Status(-1, "DuplicateToken failed");
+  }
+
+  GENERIC_MAPPING mapping = {static_cast<ACCESS_MASK>(-1),
+                             static_cast<ACCESS_MASK>(-1),
+                             static_cast<ACCESS_MASK>(-1),
+                             static_cast<ACCESS_MASK>(-1)};
+  PRIVILEGE_SET privileges = {0};
+
+  BOOL access_status = FALSE;
+  DWORD granted_access = 0;
+  DWORD privileges_length = sizeof(privileges);
+
+  mapping.GenericRead = FILE_GENERIC_READ;
+  mapping.GenericWrite = FILE_GENERIC_WRITE;
+  mapping.GenericExecute = FILE_GENERIC_EXECUTE;
+  mapping.GenericAll = FILE_ALL_ACCESS;
+
+  ::MapGenericMask(&access_rights, &mapping);
+
+  status = ::AccessCheck(sd,
+                         impersonate_token,
+                         access_rights,
+                         &mapping,
+                         &privileges,
+                         &privileges_length,
+                         &granted_access,
+                         &access_status);
+  ::CloseHandle(impersonate_token);
+
+  if (!status) {
+    return Status(-1, "AccessCheck failed");
+  }
+
+  if (access_status) {
+    return Status(0, "OK");
+  }
+
+  return Status(1, "Bad mode for file");
+}
+
+static Status hasAccess(const fs::path& path, mode_t mode) {
+  DWORD result = -1;
+  PSECURITY_DESCRIPTOR sd = nullptr;
+  SECURITY_INFORMATION security_info = OWNER_SECURITY_INFORMATION |
+                                       GROUP_SECURITY_INFORMATION |
+                                       DACL_SECURITY_INFORMATION;
+
+  result = ::GetNamedSecurityInfoW(path.wstring().c_str(),
+                                   SE_FILE_OBJECT,
+                                   security_info,
+                                   nullptr,
+                                   nullptr,
+                                   nullptr,
+                                   nullptr,
+                                   &sd);
+  if (result != ERROR_SUCCESS) {
+    return Status(-1, "GetNamedSecurityInfo failed: " + std::to_string(result));
+  }
+
+  auto status = checkAccessWithSD(sd, mode);
+  ::LocalFree(sd);
+
+  return status;
+}
+
+static Status hasAccess(HANDLE handle, mode_t mode) {
+  BOOL status = FALSE;
+  DWORD sd_size = 0;
+  SECURITY_INFORMATION security_info = OWNER_SECURITY_INFORMATION |
+                                       GROUP_SECURITY_INFORMATION |
+                                       DACL_SECURITY_INFORMATION;
+
+  status =
+      ::GetUserObjectSecurity(handle, &security_info, nullptr, 0, &sd_size);
+  if (status || (!status && ::GetLastError() != ERROR_INSUFFICIENT_BUFFER)) {
```

Comment:
This looks odd, are you sure this is correct? You want to return -1 if `status == TRUE`?